Migrate merchandise return view - base

JevgenijVisockij opened the pull request:
<!-----------------------------------------------------------------------------
Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows below with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop.com/1.7/contribute/contribution-guidelines/#pull-requests
------------------------------------------------------------------------------>

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | This PR migrates "merchandise return" view/edit BO page to symfony. This PR migrates only view of merchandise return **without product list or label printing** as planned in roadmap on https://github.com/PrestaShop/PrestaShop/issues/24421.
| Type?         | refacto
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | One item of https://github.com/PrestaShop/PrestaShop/issues/24421
| How to test?  | See below

## How to test

1. Enable merchandise returns option
2. Create some merchandise returns.
3. Go to `admin-dev/index.php/sell/customer-service/merchandise-return/` which is the list. Click on any of existing returns to open the page.
4. In this page, you should be able to change the 'order return state' and save it successfully. The design is not very good because the page is not complete.

⚠️ **PDF printing and product list are not included in this PR.** They will be implemented later. Only changing the state is implemented here. The design is not very good because the page is not complete. It will look so much better when next PRs are submitted and merged.

## About class and namespace naming

All new classes and namespaces in this PR are `OrderReturn` instead of `MerchandiseReturn` because that's the direction we decided to go with.

Everything codewise in legacy was called OrderReturn as well so we want to maintain codebase consistency. So OrderReturn in code = MerchandiseReturn in interface.

There will be a separate PR to change all existing code(mostly related to Merchandise Return grid) to reflect that (see roadmap https://github.com/PrestaShop/PrestaShop/issues/24421).


<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/19420)
<!-- Reviewable:end -->

JevgenijVisockij committed:
Migrating base for merchandise return view page

JevgenijVisockij committed:
Removed accidentally added change to database creator

PierreRambaud commented on src/Adapter/Form/ChoiceProvider/OrderReturnStateChoiceProvider.php:
```diff
@@ -0,0 +1,66 @@
+use OrderReturnState;
+use PrestaShop\PrestaShop\Core\Form\FormChoiceProviderInterface;
+
+class OrderReturnStateChoiceProvider implements FormChoiceProviderInterface
```
```suggestion
final class OrderReturnStateChoiceProvider implements FormChoiceProviderInterface
```

zuk3975 replied:
Is this still relevant? Recently it was suggested to remove final classes, isn't that right @Progi1984 ?
Though im afraid @JevgenijVisockij will be pissed off about this back and forth 😅

PierreRambaud commented on src/Adapter/OrderReturn/AbstractOrderReturnHandler.php:
```diff
@@ -0,0 +1,115 @@
+        }
+
+        if ($orderReturn->id !== $orderReturnId->getValue()) {
+            throw new OrderReturnNotFoundException($orderReturnId, sprintf('Merchandise return with id "%s" was not found.', $orderReturnId->getValue()));
```
```suggestion
            throw new OrderReturnNotFoundException($orderReturnId, sprintf('Merchandise return with id "%d" was not found.', $orderReturnId->getValue()));
```

PierreRambaud commented on src/Adapter/OrderReturn/AbstractOrderReturnHandler.php:
```diff
@@ -0,0 +1,115 @@
+        if ($orderReturnState->id !== $orderReturnStateId->getValue()) {
+            throw new OrderReturnOrderStateConstraintException(
+                $orderReturnStateId,
+                sprintf('Merchandise return state with id "%s" was not found.', $orderReturnStateId->getValue())
```
```suggestion
                sprintf('Merchandise return state with id "%d" was not found.', $orderReturnStateId->getValue())
```

JevgenijVisockij committed:
Fixed integration test

JevgenijVisockij committed:
Apply suggestions from code review

Small fixes from code review

Co-authored-by: GoT <PierreRambaud@users.noreply.github.com>

Julievrz commented on src/PrestaShopBundle/Controller/Admin/Sell/CustomerService/MerchandiseReturnController.php:
```diff
@@ -75,6 +83,95 @@ public function indexAction(Request $request, MerchandiseReturnFilters $filters)
+            ),
+            OrderReturnNotFoundException::class => $this->trans(
+                'Merchandise return not found.',
+                'Admin.Notifications.Error'
```
```suggestion
                'Admin.Orderscustomers.Notification'
```

Julievrz commented on src/PrestaShopBundle/Controller/Admin/Sell/CustomerService/MerchandiseReturnController.php:
```diff
@@ -75,6 +83,95 @@ public function indexAction(Request $request, MerchandiseReturnFilters $filters)
+            ],
+            MissingOrderReturnRequiredFieldsException::class => $this->trans(
+                'Missing required fields for merchandise return.',
+                'Admin.Notifications.Error'
```
```suggestion
                'Admin.OrdersCustomers.Notification'
```

Julievrz commented on src/PrestaShopBundle/Controller/Admin/Sell/CustomerService/MerchandiseReturnController.php:
```diff
@@ -75,6 +83,95 @@ public function indexAction(Request $request, MerchandiseReturnFilters $filters)
+            ],
+            UpdateOrderReturnException::class => $this->trans(
+                'An error occurred while trying to update merchandise return.',
+                'Admin.Notifications.Error'
```
```suggestion
                'Admin.OrdersCustomers.Notification'
```

JevgenijVisockij committed:
Apply suggestions from code review

Wording domain changes

Co-authored-by: Julie Varisellaz <70583503+Julievrz@users.noreply.github.com>

JevgenijVisockij committed:
Added cache clearing and removed unneeded step

Progi1984 commented on tests/Integration/Behaviour/Features/Context/Domain/OrderReturnFeatureContext.php:
```diff
@@ -0,0 +1,132 @@
+        $orderReturn->id_customer = (int) $order->id_customer;
+        $orderReturn->id_order = $order->id;
+        $orderReturn->question = 'Why?';
+
```
```suggestion
```

Progi1984 commented on tests/Integration/Behaviour/Features/Context/Domain/OrderReturnFeatureContext.php:
```diff
@@ -0,0 +1,132 @@
+
+        $orderReturn->state = 1;
+        $orderReturn->add();
+        $orderDetailIds = [];
+        $quantities = [];
+        $customizationIds = [];
+        $customizationQuantityIds = [];
```
```suggestion
        $quantities = $customizationIds = $customizationQuantityIds = [];
```

Progi1984 replied:
```suggestion
        $orderDetailIds = $quantities = $customizationIds = $customizationQuantityIds = [];
```

Progi1984 commented on src/Core/Form/IdentifiableObject/DataProvider/OrderReturnFormDataProvider.php:
```diff
@@ -0,0 +1,72 @@
+/**
+ * Provides data for order return edit form
+ */
+final class OrderReturnFormDataProvider implements FormDataProviderInterface
```
```suggestion
class OrderReturnFormDataProvider implements FormDataProviderInterface
```

Progi1984 commented on src/Core/Form/IdentifiableObject/DataHandler/OrderReturnFormDataHandler.php:
```diff
@@ -0,0 +1,70 @@
+/**
+ * Saves or updates order return data submitted in form
+ */
+final class OrderReturnFormDataHandler implements FormDataHandlerInterface
```
```suggestion
class OrderReturnFormDataHandler implements FormDataHandlerInterface
```

Progi1984 commented on src/Core/Domain/OrderReturn/ValueObject/OrderReturnId.php:
```diff
@@ -0,0 +1,75 @@
+     */
+    public function __construct(int $id)
+    {
+        $this->assertIsIntegerGreaterThanZero($id);
```
Could you unit test this behavior, please ?

Progi1984 commented on src/Core/Domain/OrderReturn/QueryResult/EditableOrderReturn.php:
```diff
@@ -0,0 +1,178 @@
+     */
+    public function getCustomerFullName(): string
+    {
+        return sprintf('%s %s', $this->customerFirstName, $this->customerLastName);
```
Could you unit test this behavior ?

Progi1984 commented on src/Adapter/OrderReturn/QueryHandler/GetOrderReturnForEditingHandler.php:
```diff
@@ -0,0 +1,65 @@
+/**
+ * Handles query which gets order return for editing
+ */
+final class GetOrderReturnForEditingHandler extends AbstractOrderReturnHandler implements GetOrderReturnForEditingHandlerInterface
```
```suggestion
class GetOrderReturnForEditingHandler extends AbstractOrderReturnHandler implements GetOrderReturnForEditingHandlerInterface
```

Progi1984 commented on src/Adapter/OrderReturn/CommandHandler/UpdateOrderReturnStateHandler.php:
```diff
@@ -0,0 +1,78 @@
+        $this->assertRequiredFieldsAreNotMissing($orderReturn);
+
+        if (false === $orderReturn->validateFields(false)) {
+            throw new OrderReturnException('Order return contains invalid field values');
```
Could you add behat tests for this behavior ?

JevgenijVisockij replied:
Not really.

It's impossible to get here with the current code structure. Only state is updated, and it will always be int due to strict types. Any int will always pass validate fields for state.

This here just in case something goes really wrong in the code somewhere else and affects this handler.

I think I could add unit test, where updateOrderReturnWithCommandData is mocked to alert state and make it incorrect. What do you think?

zuk3975 replied:
Validation could be improved by validating each field using legacy object model validator (as done in Product subdomains). But this doesnt block the pr i guess, it could be done in separate one

Progi1984 reviewed (commented):
Some feedbacks

JevgenijVisockij committed:
Update src/Adapter/OrderReturn/QueryHandler/GetOrderReturnForEditingHandler.php

Removed final from handler

Co-authored-by: Progi1984 <progi1984@gmail.com>

JevgenijVisockij committed:
Apply suggestions from code review

Removed some uneeded final classes, also some uneeded spaces

Co-authored-by: Progi1984 <progi1984@gmail.com>

JevgenijVisockij committed:
ADded unit tests for merchandise return

khouloudbelguith commented:
Hi @JevgenijVisockij,

I have two issues:

**Issue1**
1. Go to BO > Customer Service > Merchandise Returns 
2. Edit a Return
3. See error in the console

![image](https://user-images.githubusercontent.com/16067358/123629118-2cf2fe80-d814-11eb-8c65-0cbcb7bf8e9a.png)
![image](https://user-images.githubusercontent.com/16067358/123629310-6e83a980-d814-11eb-90d4-f72b8a3f2740.png)

**Issue2**
1. Create a return from the FO with a customer explanation that contains special characters 
![image](https://user-images.githubusercontent.com/16067358/123629569-b73b6280-d814-11eb-9687-82b1c83645a9.png)
2. Go to BO > Customer Service > Merchandise Returns
3. Edit the last Return
4. See error 
![image](https://user-images.githubusercontent.com/16067358/123629847-11d4be80-d815-11eb-9cc9-57d336df257d.png)

Legacy page => ok
![image](https://user-images.githubusercontent.com/16067358/123629942-30d35080-d815-11eb-8e6a-69815330506d.png)


Thank you!

JevgenijVisockij committed:
Fixed issue with symbols < not showing as escaped values in order return customer explanation.

JevgenijVisockij committed:
Removed no longer needed javascript

JevgenijVisockij committed:
Fixed issue with wrong exception thrown when order return is not found

JevgenijVisockij committed:
EmployeeController.php

JevgenijVisockij committed:
Moved order return test after order

# Conflicts:
#	tests/Integration/Behaviour/behat.yml

JevgenijVisockij committed:
Fixed order return test

JevgenijVisockij commented:
> Hi @JevgenijVisockij,
> 
> I have two issues:
> 
> **Issue1**
> 
>     1. Go to BO > Customer Service > Merchandise Returns
> 
>     2. Edit a Return
> 
>     3. See error in the console
> 
> 
> ![image](https://user-images.githubusercontent.com/16067358/123629118-2cf2fe80-d814-11eb-8c65-0cbcb7bf8e9a.png)
> ![image](https://user-images.githubusercontent.com/16067358/123629310-6e83a980-d814-11eb-90d4-f72b8a3f2740.png)
> 
> **Issue2**
> 
>     1. Create a return from the FO with a customer explanation that contains special characters
>        ![image](https://user-images.githubusercontent.com/16067358/123629569-b73b6280-d814-11eb-9687-82b1c83645a9.png)
> 
>     2. Go to BO > Customer Service > Merchandise Returns
> 
>     3. Edit the last Return
> 
>     4. See error
>        ![image](https://user-images.githubusercontent.com/16067358/123629847-11d4be80-d815-11eb-9cc9-57d336df257d.png)
> 
> 
> Legacy page => ok
> ![image](https://user-images.githubusercontent.com/16067358/123629942-30d35080-d815-11eb-8e6a-69815330506d.png)
> 
> Thank you!

Hey. Thanks for Q&A! Fixed both issues.

zuk3975 commented on src/Adapter/OrderReturn/AbstractOrderReturnHandler.php:
```diff
@@ -0,0 +1,115 @@
+/**
+ * Provides reusable methods for order return command/query handlers
+ */
+abstract class AbstractOrderReturnHandler extends AbstractObjectModelHandler
```
Ah this is also kinda outdated, i dont know if we should require to change it or not. But instead of abstract handler there should be services. Like to get orderReturn or orderReturnState there should be related repository injected and for asserrion maybe a dedicated validator service. There are plenty examples in product domain

jolelievre replied:
I agree it would be preferable if we use the same architecture from now on Sorry for the extra work to convert your abstract classes into dedicated services But I promise in the short term it will make your code base easier to maintain, understand and reuse

You can find examples in the codebase and you even have this documentation that was written https://devdocs.prestashop.com/1.7/development/architecture/domain/domain-services/

Although it may not be complete enough, feel free to tell us if something is missing

zuk3975 commented on src/Adapter/OrderReturn/CommandHandler/UpdateOrderReturnStateHandler.php:
```diff
@@ -0,0 +1,78 @@
+            throw new OrderReturnException('Order return contains invalid field values');
+        }
+
+        if (false === $orderReturn->update()) {
```
Object model write operations like add/update/save should be wrapped in try-catch to disallow old PrestaShopException bubbling up. Once again this is also covered in AbstractObjectModelRepository that you could use

jolelievre replied:
The getter function also have automatic check of the object existence by returning some exception when they are not found So it should make the whole behaviour more robust as it asserts many things "out of the box"

Well, not everything of course but if you follow the initial architecture and abstract classes it does most of it anyway ^^

zuk3975 commented on src/Core/Domain/OrderReturn/Command/UpdateOrderReturnStateCommand.php:
```diff
@@ -0,0 +1,77 @@
+
+    /**
+     * @var OrderReturnStateId|null
+     */
```
Cannot be null as i see in construct

zuk3975 commented on src/Core/Domain/OrderReturn/Exception/MissingOrderReturnRequiredFieldsException.php:
```diff
@@ -0,0 +1,64 @@
+/**
+ * Is thrown when some of the required fields are missing when saving order return
+ */
+class MissingOrderReturnRequiredFieldsException extends OrderReturnException
```
Is it somehow different than any {domain}ConstraintException with dedicated codes? Couldnt it be implemented as everywhere else? If some field is missing, then OrderReturnConstraintException::{SOME_SPECIFIC_FIELD_IS_INVALID} should be thrown, no?

jolelievre replied:
I guess the main difference here is that it contains the list of missing fields So a dedicated exception is required I guess

Maybe, at worst, it could extend `OrderReturnConstraintException` instead of `OrderReturnException` if we do consider this is a constraint

zuk3975 replied:
not sure if Im missing something, but if we use ObjectModelValidator and manually validate each field inside, then we have an error code for each field and its never a list of missing fields. (like there is never a list of missing fields in product - instead we have one error which represents one invalid field - either its missing or invalid in any other way)

JevgenijVisockij replied:
Yeah I don't think this exception is used anymore so I will delete it

zuk3975 commented on src/Core/Domain/OrderReturn/QueryHandler/GetOrderReturnForEditingHandlerInterface.php:
```diff
@@ -0,0 +1,43 @@
+     *
+     * @return EditableOrderReturn
+     */
+    public function handle(GetOrderReturnForEditing $query): EditableOrderReturn;
```
Just for consistency- i thin we call most of these 'somethimgForEditing', not EditableSomething, but not sure. And its not in docs I guess so i doubt its required 🤷‍♂️

jolelievre replied:
Hmm actually we did this for product queries like here https://github.com/PrestaShop/PrestaShop/blob/0b6748e4770593e753cde8fdc3e88c1f61a76092/src/Core/Domain/Product/QueryHandler/GetProductForEditingHandlerInterface.php#L42

(and most of our other queries)

But historically seems like we were more into the `EditableThing`
https://github.com/PrestaShop/PrestaShop/blob/3d06748975e00c548e940bf82862d9d211ae94d4/src/Core/Domain/Contact/QueryHandler/GetContactForEditingHandlerInterface.php#L40
https://github.com/PrestaShop/PrestaShop/blob/3d06748975e00c548e940bf82862d9d211ae94d4/src/Core/Domain/Tax/QueryHandler/GetTaxForEditingHandlerInterface.php#L40

Seems like we started using the `ForEditing` `ForViewing` with order queries (but I'm not even sure as there seems to be a mix)

JevgenijVisockij replied:
I am pretty sure I just copied idea from one of other classes. But if we are moving to use  `ForEditing` instead of `Editable` that's fine for me. I guess `ForEditing` at least uses the same naming as handler.

zuk3975 commented on src/Core/Domain/OrderReturn/ValueObject/OrderReturnId.php:
```diff
@@ -0,0 +1,75 @@
+    private function assertIsIntegerGreaterThanZero(int $value): void
+    {
+        if (0 >= $value) {
+            throw new OrderReturnConstraintException(sprintf('Invalid order return id "%s".', $value));
```
Missing exception code INVALID_ID I think

zuk3975 commented on src/Core/Form/IdentifiableObject/DataHandler/OrderReturnFormDataHandler.php:
```diff
@@ -0,0 +1,70 @@
+    /**
+     * @var CommandBusInterface
+     */
+    private $bus;
```
We have 2 buses. It would be better to call it explicitly - $commandBus (just a suggestion though)

jolelievre replied:
To tell `commandBus` apart from the `queryBus` indeed (although actually, they are the same bus 😅 ) 

https://github.com/PrestaShop/PrestaShop/blob/0247232e2b0a6670add7c19ef60ec9349d8438a5/src/PrestaShopBundle/Resources/config/services/core/command_bus.yml#L10-L12

And we sometimes used `bus` alone as well ^^

https://github.com/PrestaShop/PrestaShop/blob/9e427e3e638e2c460f7ab0b0e0baf0dc685bfcdf/src/Core/Form/IdentifiableObject/DataHandler/ProductFormDataHandler.php#L44

But I agree as a suggestion

JevgenijVisockij replied:
Makes sense, thanks!

zuk3975 commented on src/Core/Form/IdentifiableObject/DataHandler/OrderReturnFormDataHandler.php:
```diff
@@ -0,0 +1,70 @@
+     * @param array $data
+     */
+    public function create(array $data): void
+    {
```
It is time to decide if we trigger a warning in such empty methods (there are quite a couple of them already). @jolelievre @PierreRambaud (I think you've suggested it in one of my PRs) Wdyt?

jolelievre replied:
I believe I usually just add a comment 😅
But why not using a trigger warning yes! Like we do with deprecations

```php
    @trigger_error(OrderReturn::class . ' is not created by form, this method should never be called', E_USER_WARNING);
```

jolelievre commented on src/Adapter/Form/ChoiceProvider/OrderReturnStateChoiceProvider.php:
```diff
@@ -0,0 +1,66 @@
+        $orderStates = OrderReturnState::getOrderReturnStates($this->contextLangId);
+
+        foreach ($orderStates as $orderState) {
+            $choices[$orderState['name']] = (int) $orderState['id_order_return_state'];
```
I know we had some bugs related to Order status when they have duplicated names I guess the same problem could happen here

@Progi1984 If I remember well you are the one who fixed the issue for orders' status, what was the strategy used?

matks replied:
If we dont know how to do it better, I would suggest 2 strategies:

1. "Crash early"
```
if (array_key_exists($orderState['name'], $choices) {
    throw new \RuntimeException('This should NEVER HAPPEN !');
}
```

2. Silent resilience
```
if (array_key_exists($orderState['name'], $choices) {
    $this->logThereIsDuplicateOrderStateName();
    continue;
}
```

JevgenijVisockij replied:
Hmm I added first case. But it potentially can happen, because there is no restricction on adding same order state names.

jolelievre commented on src/Adapter/OrderReturn/QueryHandler/GetOrderReturnForEditingHandler.php:
```diff
@@ -0,0 +1,65 @@
+/**
+ * Handles query which gets order return for editing
+ */
+class GetOrderReturnForEditingHandler extends AbstractOrderReturnHandler implements GetOrderReturnForEditingHandlerInterface
```
Here you will probably need to create two additional repositories:
- OrderRepository
- CustomerRepository

It will be useful for you for future handlers anyway, and relying on the abstract classes it's relatively fast to add them

jolelievre commented on src/Core/Domain/OrderReturn/Exception/OrderReturnNotFoundException.php:
```diff
@@ -0,0 +1,66 @@
+    /**
+     * @var OrderReturnId
+     */
+    private $orderReturnId;
```
I'm not sure you need to store the ID here (since it is not used by the controller when the error is displayed)

Besides if you implement the Repository using our usual abstract class it only handles a type of exception but we never planned to inject the ID in it, so you can simplify this exception I think

That said maybe we should have handled an ID parameter, I guess it's possible to handle it in the abstract repository relying on a common interface that would have a `setEntityID/getEntityId` getters/setters

JevgenijVisockij replied:
I think exception having an id of what it did not find is a good idea, because if exception happens when multiple objects are handled usually the first question is "Which ID was not found?"

But you are right I don't use it here, so I will remove it for now. I will take a look at abstract repository for a way to include id.

jolelievre commented on src/PrestaShopBundle/Controller/Admin/Sell/CustomerService/MerchandiseReturnController.php:
```diff
@@ -75,6 +83,95 @@ public function indexAction(Request $request, MerchandiseReturnFilters $filters)
+
+            $result = $formHandler->handleFor($orderReturnId, $form);
+
+            if ($result->getIdentifiableObjectId()) {
```
```suggestion
            if ($result->isSubmitted() && $result->isValid()) {
```

We don't care much about the returned ID here since it's an edit action, the most important is that the updated suceeded

jolelievre commented on src/PrestaShopBundle/Controller/Admin/Sell/CustomerService/MerchandiseReturnController.php:
```diff
@@ -75,6 +83,95 @@ public function indexAction(Request $request, MerchandiseReturnFilters $filters)
+        return $this->render('@PrestaShop/Admin/Sell/CustomerService/OrderReturn/edit.html.twig', [
+            'help_link' => $this->generateSidebarLink($request->attributes->get('_legacy_controller')),
+            'enableSidebar' => true,
+            'layoutTitle' => sprintf($this->trans('Return Merchandise Authorization (RMA)', 'Admin.Orderscustomers.Feature')),
```
Why do you need a `sprintf` here? You have no argument to inject, and even if you did `trans` can handle that

JevgenijVisockij replied:
No idea why it's there. Nice catch!

jolelievre commented on src/PrestaShopBundle/Form/Admin/Sell/CustomerService/MerchandiseReturn/OrderReturnFormDataProvider.php:
```diff
@@ -0,0 +1,69 @@
+/**
+ * Provides order return options form with data
+ */
+final class OrderReturnFormDataProvider implements FormDataProviderInterface
```
Mmm this one is weird it's confusing with the `PrestaShop\PrestaShop\Core\Form\IdentifiableObject\DataProvider\OrderReturnFormDataProvider` and besides I don't see any uses of this service anywhere

It more looks more like a provider for a Shop configuration form

If you need a service to change the options of the `OrderReturnType` you can instead rely on the `FormOptionsProviderInterface` The only use of this interface for now is this one https://github.com/PrestaShop/PrestaShop/blob/7122c895b442ab8c4bb83aee3a38fbb9ece1c8b5/src/Core/Form/IdentifiableObject/OptionProvider/CombinationFormOptionsProvider.php

jolelievre commented on src/PrestaShopBundle/Resources/config/routing/admin/sell/customer_service/merchandise_return.yml:
```diff
@@ -23,3 +23,27 @@ admin_merchandise_returns_filter:
+    requirements:
+        orderReturnId: \d+
+
+admin_order_returns_update:
```
This route is a duplicate of `admin_order_returns_edit` I don't understand its purpose

jolelievre replied:
Oh I get it, it's because one is GET and the other is POST, if you had two distinct actions in your controller why not But I think you can fuse them and allow `[GET, POST]` as methods

jolelievre commented on src/PrestaShopBundle/Resources/config/services/core/form/form_data_handler.yml:
```diff
@@ -118,6 +118,13 @@ services:
+    class: 'PrestaShop\PrestaShop\Core\Form\IdentifiableObject\DataHandler\OrderReturnFormDataHandler'
+    arguments:
+      - '@prestashop.core.command_bus'
+      - '@=service("prestashop.adapter.legacy.context").getContext().shop.id'
```
Why injecting those two extra parameters? They are not used in the `OrderReturnFormDataHandler` class

JevgenijVisockij replied:
Yeah they shouldn't be needed. Will remove them!

jolelievre commented on src/PrestaShopBundle/Resources/config/services/core/form/form_handler.yml:
```diff
@@ -124,6 +124,12 @@ services:
 
+  prestashop.core.form.identifiable_object.handler.order_return_form_handler:
+    class: 'PrestaShop\PrestaShop\Core\Form\IdentifiableObject\Handler\FormHandler'
+    factory: 'prestashop.core.form.identifiable_object.handler.form_handler_factory:create'
```
```suggestion
    factory: [ '@prestashop.core.form.identifiable_object.handler.form_handler_factory, 'create']
```

Since we are now using Symfony 4.4, you should use the new syntax for factories https://symfony.com/doc/4.4/service_container/factories.html#non-static-factories

Since this PR has been merged https://github.com/PrestaShop/PrestaShop/pull/25052 I believe this is the right way now, right @atomiix ?

jolelievre commented on src/PrestaShopBundle/Resources/config/routing/admin/sell/customer_service/merchandise_return.yml:
```diff
@@ -23,3 +23,27 @@ admin_merchandise_returns_filter:
+    path: /{orderReturnId}/edit
+    methods: [GET]
+    defaults:
+        _controller: PrestaShopBundle:Admin/Sell/CustomerService/MerchandiseReturn:edit
```
```suggestion
        _controller: PrestaShopBundle\Controller\Admin\Sell\CustomerService\MerchandiseReturnController::editAction
```

We must use full notation now since this PR has been merged https://github.com/PrestaShop/PrestaShop/pull/25052, for consistency but mostly because short notation will no longer be supported in the future

jolelievre commented on src/PrestaShopBundle/Resources/views/Admin/Sell/CustomerService/OrderReturn/Blocks/form.html.twig:
```diff
@@ -0,0 +1,92 @@
+    {{ 'Merchandise returns'|trans({}, 'Admin.Orderscustomers.Feature') }}
+  </h3>
+
+  <div class="card-block row">
```
The integration of the whole form seems a bit "complex" in the sense that you duplicate a lot of templates from the form theme manually When a form should mostly be rendered with a simple `form_row(form)`

jolelievre commented on src/PrestaShopBundle/Resources/views/Admin/Sell/CustomerService/OrderReturn/Blocks/form.html.twig:
```diff
@@ -0,0 +1,92 @@
+        </div>
+      </div>
+
+      <div class="form-group row">
```
This field for example could be integrated in the `OrderReturnFormType` and use this kind of form type
https://github.com/PrestaShop/PrestaShop/blob/720ff9fef6f59a8becb3db6225cf4d505ac1a1ee/src/PrestaShopBundle/Form/Admin/Type/TextPreviewType.php

This is a new form type I introduced in this PR https://github.com/PrestaShop/PrestaShop/pull/24878 it hasn't been merged yet, but it answers the need you have here "I want to display some value in a form without it being editable"

jolelievre replied:
I also needed a https://github.com/PrestaShop/PrestaShop/blob/720ff9fef6f59a8becb3db6225cf4d505ac1a1ee/src/PrestaShopBundle/Form/Admin/Type/ImagePreviewType.php for images

But maybe in your case you would need a `LinkType` or `LinkPreviewType` (for the two previous form fields: order view and customer view)

JevgenijVisockij replied:
Good idea about LinkPreviewType! I will refactor the form

jolelievre commented on tests/Integration/Behaviour/Features/Context/Domain/OrderReturnFeatureContext.php:
```diff
@@ -0,0 +1,128 @@
+     * @throws \PrestaShopDatabaseException
+     * @throws \PrestaShopException
+     */
+    public function createOrderReturnState(string $orderReturnStateReference): void
```
Since these statements are mainly used to created fixtures for your behat tests I don't think they should be in this Domain context
Usually we separate these context in the "upper" namespace, a bit like we did for Order:

Context to manipulate legacy objects directly (for features or low levels checks)
https://github.com/PrestaShop/PrestaShop/blob/ee00c133ac5140a427f45b6ea5d948cb34f9123b/tests/Integration/Behaviour/Features/Context/OrderFeatureContext.php 

Context that uses CQRS domain commands:
https://github.com/PrestaShop/PrestaShop/blob/22d5e6b339f481423c89678c065f5681833c11f8/tests/Integration/Behaviour/Features/Context/Domain/OrderFeatureContext.php

jolelievre replied:
So `createOrderReturnState` and `createOrderReturn` should be moved in a different context But you can still use `SharedStorage::getStorage()` as a common "proxy"

jolelievre commented on tests/Integration/Behaviour/Features/Context/Domain/OrderReturnFeatureContext.php:
```diff
@@ -0,0 +1,128 @@
+     */
+    public function updateOrderReturnState(string $orderReturnReference, string $orderRetunStateReference): void
+    {
+        $orderReturnId = SharedStorage::getStorage()->get($orderReturnReference);
```
```suggestion
        $orderReturnId = $this->getSharedStorage()->get($orderReturnReference);
```
Since you extend `AbstractDomainFeatureContext` here you can use this getter

jolelievre commented on tests/Integration/Behaviour/Features/Context/Domain/OrderReturnFeatureContext.php:
```diff
@@ -0,0 +1,128 @@
+     * @throws PrestaShopDatabaseException
+     * @throws PrestaShopException
+     */
+    public function checkOrderReturnState(string $orderReturnReference, string $orderReturnStateReference): void
```
Either you should move this to the "legacy" behat context Or you can keep it here but then you should rely on the `GetOrderReturnForEditing` which would be a good thing because it would allow you to also test the `GetOrderReturnForEditingHandler` this way

JevgenijVisockij replied:
Good point. Will refactor to use order return for editing.

jolelievre commented on tests/Integration/Behaviour/Features/Context/Domain/OrderReturnFeatureContext.php:
```diff
@@ -0,0 +1,128 @@
+        $orderReturn->id_customer = (int) $order->id_customer;
+        $orderReturn->id_order = $order->id;
+        $orderReturn->question = 'Why?';
+        $orderReturn->state = 1;
```
This refers to an `OrderState` in database right? Maybe you should rather use a state you created via a reference stored in the SharedStorage?

It's just that the hardcode 1 value seems a bit arbitrary, as we don't even know which state it relates to or if it even exists

JevgenijVisockij replied:
I have state as 1 because first state doesn't really matter. It might not even exist as far as test is concerned.

What I am testing for is that it correctly changes to state referenced in storage after update.

jolelievre commented on tests/Integration/Behaviour/Features/Scenario/OrderReturn/order_return_management.feature:
```diff
@@ -0,0 +1,27 @@
+      | status              | Awaiting bank wire payment |
+    And I add order return state "orderReturnState1"
+    And I add order return "testOrderReturn" from order "bo_order1"
+  Scenario: Change order return status
```
```suggestion

  Scenario: Change order return status
```

Just add a line break between each Background and Scenarios blocks so we can easily read and identify them

JevgenijVisockij committed:
Changed order state update logic to use domain services

JevgenijVisockij committed:
Changed order return update to use customer and order repositories instead of using legacy objects outright.

Renamed EditableOrderReturn and EditableOrderReturnState into OrderReturnForEditing and OrderReturnStateForEditing

JevgenijVisockij committed:
Fixed error codes, added error code to OrderReturnId exception. Added warning to OrderReturn create method

JevgenijVisockij committed:
Update src/PrestaShopBundle/Controller/Admin/Sell/CustomerService/MerchandiseReturnController.php

Changed from checknig for id instead to check for isValid and is submitted when saving order return

Co-authored-by: Jonathan Lelievre <jo.lelievre@gmail.com>

JevgenijVisockij committed:
Update src/PrestaShopBundle/Resources/config/routing/admin/sell/customer_service/merchandise_return.yml

Using full notation to avoid compatability problems

Co-authored-by: Jonathan Lelievre <jo.lelievre@gmail.com>

JevgenijVisockij committed:
Removed some uneeded things like sprintf or DataProvider

JevgenijVisockij committed:
Update src/PrestaShopBundle/Resources/config/services/core/form/form_handler.yml

Using Symfony 4.4 syntax for factory

Co-authored-by: Jonathan Lelievre <jo.lelievre@gmail.com>

JevgenijVisockij committed:
Update tests/Integration/Behaviour/Features/Context/Domain/OrderReturnFeatureContext.php

Using getter rather then static function

Co-authored-by: Jonathan Lelievre <jo.lelievre@gmail.com>

JevgenijVisockij committed:
Update tests/Integration/Behaviour/Features/Scenario/OrderReturn/order_return_management.feature

Adding line break for clarity

Co-authored-by: Jonathan Lelievre <jo.lelievre@gmail.com>

JevgenijVisockij committed:
Refactored test to use order return for editing

JevgenijVisockij committed:
Renamed order return state back as it shouldn't part of this PR

JevgenijVisockij committed:
Renamed order return state back as it shouldn't part of this PR

JevgenijVisockij committed:
Fixed factory definition in services and removed no longer used path in forms save

Progi1984 commented on src/PrestaShopBundle/Resources/config/services/adapter/order_return.yml:
```diff
@@ -0,0 +1,34 @@
+    _defaults:
+        public: true
+
+    prestashop.adapter.order_return.query_handler.get_order_return_for_editing:
```
```suggestion
    prestashop.adapter.order_return.query_handler.get_order_return_for_editing_handler:
```

Progi1984 commented on src/PrestaShopBundle/Resources/config/services/adapter/order_return.yml:
```diff
@@ -0,0 +1,34 @@
+           - '@prestashop.core.domain.customer.repository.customer_repository'
+           - '@prestashop.core.domain.order.repository.order_repository'
+
+    prestashop.adapter.order_return.command_handler.edit_order_return_handler:
```
```suggestion
    prestashop.adapter.order_return.command_handler.edit_order_return_state_handler:
```

Progi1984 commented on src/PrestaShopBundle/Controller/Admin/Sell/CustomerService/MerchandiseReturnController.php:
```diff
@@ -75,6 +82,91 @@ public function indexAction(Request $request, MerchandiseReturnFilters $filters)
+     *
+     * @return array
+     */
+    private function getErrorMessages(Exception $e = null): array
```
```suggestion
    private function getErrorMessages(): array
```

Progi1984 commented on src/Core/Domain/Order/Repository/OrderRepository.php:
```diff
@@ -0,0 +1,55 @@
+    public function get(OrderId $orderId): Order
+    {
+        /** @var Order $order */
+        $order = $this->getObjectModel($orderId->getValue(), Order::class, OrderNotFoundException::class);
+
+        return $order;
```
```suggestion
        return $this->getObjectModel($orderId->getValue(), Order::class, OrderNotFoundException::class);
```

Could you add integration test for checking the exception ?

Progi1984 commented on src/Core/Domain/Order/Repository/OrderRepository.php:
```diff
@@ -0,0 +1,55 @@
+     *
+     * @return Order
+     *
+     * @throws CoreException
```
```suggestion
     * @throws OrderNotFoundException
```

Progi1984 commented on src/Core/Domain/Customer/Repository/CustomerRepository.php:
```diff
@@ -0,0 +1,55 @@
+    public function get(CustomerId $customerId): Customer
+    {
+        /** @var Customer $customer */
+        $customer = $this->getObjectModel($customerId->getValue(), Customer::class, CustomerNotFoundException::class);
+
+        return $customer;
```
```suggestion
        return $this->getObjectModel($customerId->getValue(), Customer::class, CustomerNotFoundException::class);
```
Could you add integration test for checking the exception ?

Progi1984 commented on src/Adapter/OrderReturn/Repository/OrderReturnStateRepository.php:
```diff
@@ -0,0 +1,66 @@
+     *
+     * @return OrderReturnState
+     *
+     * @throws OrderReturnException
```
```suggestion
     * @throws OrderReturnException
     * @throws OrderReturnConstraintException
```

Progi1984 commented on src/Adapter/OrderReturn/Repository/OrderReturnRepository.php:
```diff
@@ -0,0 +1,81 @@
+    public function get(OrderReturnId $orderReturnId): OrderReturn
+    {
+        /** @var OrderReturn $orderReturn */
+        $orderReturn = $this->getObjectModel($orderReturnId->getValue(), OrderReturn::class, OrderReturnNotFoundException::class);
+
+        return $orderReturn;
```
```suggestion
        return $this->getObjectModel($orderReturnId->getValue(), OrderReturn::class, OrderReturnNotFoundException::class);
```

Could you add integration test for checking the exception ?

Progi1984 reviewed (commented):
Some feedbacks

JevgenijVisockij committed:
Updated the form to use preview form types

JevgenijVisockij committed:
Apply suggestions from code review

Suggestions for exception comments

Co-authored-by: Progi1984 <progi1984@gmail.com>

JevgenijVisockij committed:
Apply suggestions from code review

Corrected service names

Co-authored-by: Progi1984 <progi1984@gmail.com>

JevgenijVisockij committed:
Moved order return and order return state creation to a different context

JevgenijVisockij committed:
CS fixer fixes

Progi1984 commented on src/PrestaShopBundle/Form/Admin/Sell/CustomerService/MerchandiseReturn/OrderReturnType.php:
```diff
@@ -0,0 +1,105 @@
+            ])
+            ->add('order', TextPreviewType::class, [
+                'label' => $this->trans('Order', 'Admin.Global'),
+                'attr' => [
+                    'class' => 'form-control-value',
+                ],
+                'row_attr' => [
+                    'class' => 'form-control-value',
+                ],
```
`attr` & `row_attr` ?

Progi1984 reviewed (commented):
Some missing feedbacks

JevgenijVisockij committed:
Refactored repositories to work properly added integration tests for them.

JevgenijVisockij committed:
Added test for order return data provider

JevgenijVisockij committed:
CS Fixer fixes

JevgenijVisockij committed:
Apply suggestions from code review

Made OrderReturn::get smaller

Co-authored-by: Progi1984 <progi1984@gmail.com>

Progi1984 commented on src/Adapter/Customer/Repository/CustomerRepository.php:
```diff
@@ -0,0 +1,72 @@
+     *
+     * @return Customer
+     *
+     * @throws CoreException
```
```suggestion
     * @throws CustomerNotFoundException
     * @throws CoreException
```

Progi1984 commented on tests/Unit/Core/Form/IdentifiableObject/DataProvider/OrderReturnFormDataProviderTest.php:
```diff
@@ -0,0 +1,151 @@
+     *
+     * @return MockObject|CommandBusInterface
+     */
+    private function createQueryBusMock(OrderReturnForEditing $orderReturnForEditing)
```
```suggestion
    private function createQueryBusMock(OrderReturnForEditing $orderReturnForEditing): CommandBusInterface
```

Progi1984 commented on src/PrestaShopBundle/Form/Admin/Sell/CustomerService/MerchandiseReturn/OrderReturnType.php:
```diff
@@ -0,0 +1,109 @@
+     */
+    public function __construct(
+        TranslatorInterface $translator,
+        array $locales,
```
Missing PHPDoc

Progi1984 commented on src/Adapter/OrderReturn/Repository/OrderReturnRepository.php:
```diff
@@ -0,0 +1,79 @@
+     */
+    public function get(OrderReturnId $orderReturnId): OrderReturn
+    {
+        /** @var OrderReturn $orderReturn */
+        return $this->getObjectModel($orderReturnId->getValue(), OrderReturn::class, OrderReturnNotFoundException::class);
```
```suggestion
        /** @var OrderReturn $orderReturn */
        $orderReturn = $this->getObjectModel($orderReturnId->getValue(), OrderReturn::class, OrderReturnNotFoundException::class);
        return $orderReturn;
```

Sorry for the revert but PHPStan doesn't like `getObjectModel`

Progi1984 commented on tests/Integration/Adapter/Order/Repository/OrderRepositoryTest.php:
```diff
@@ -0,0 +1,55 @@
+        $this->orderRepository = self::$container->get('prestashop.adapter.order.repository.order_repository');
+    }
+
+    public function testGetThrowsException()
```
```suggestion
    public function testGetThrowsException(): void
```

Progi1984 commented on src/Adapter/Order/Repository/OrderRepository.php:
```diff
@@ -0,0 +1,72 @@
+     *
+     * @return Order
+     *
+     * @throws CoreException
```
```suggestion
     * @throws CoreException
     * @throws OrderNotFoundException
```

Progi1984 commented on tests/Integration/Adapter/Customer/Repository/CustomerRepositoryTest.php:
```diff
@@ -0,0 +1,55 @@
+        $this->customerRepository = self::$container->get('prestashop.adapter.customer.repository.customer_repository');
+    }
+
+    public function testGetThrowsException()
```
```suggestion
    public function testGetThrowsException(): void
```

Progi1984 reviewed (commented):
Some small fixes and it's ok for me

matks commented on src/Adapter/Customer/Repository/CustomerRepository.php:
```diff
@@ -0,0 +1,72 @@
+<?php
+/**
+ * 2007-2020 PrestaShop SA and Contributors
```
"Old style" license headers 😝  we now use `Copyright since 2007 PrestaShop SA and Contributors`

JevgenijVisockij replied:
Riight. Must have copied them from somewhere. Will fix!

matks commented on src/Adapter/OrderReturn/Validator/OrderReturnValidator.php:
```diff
@@ -0,0 +1,70 @@
+     *
+     * @throws OrderReturnConstraintException
+     */
+    public function validate(OrderReturn $orderReturn): void
```
Clean and efficient 👍

matks commented on src/Core/Domain/OrderReturn/Exception/OrderReturnConstraintException.php:
```diff
@@ -0,0 +1,70 @@
+    /**
+     * When date updated is not valid
+     */
+    public const INVALID_DATE_UPD = 70;
```
I feel like reading Java.

And Java is good ! 💪

matks commented on src/PrestaShopBundle/Form/Admin/Type/LinkPreviewType.php:
```diff
@@ -0,0 +1,74 @@
+            'error_bubbling' => true,
+            'compound' => false,
+            'required' => false,
+            'button_label' => '', // Start automatically TinyMCE
```
📖 👍

matks reviewed (dismissed):
Good for me. Only change needed on my side 😄 : the license headers

JevgenijVisockij committed:
Fixed order return context namespace

JevgenijVisockij committed:
CS fixer fixes

JevgenijVisockij committed:
Fixed copyright

JevgenijVisockij committed:
Apply suggestions from code review

Minor fixes: return types and docs

Co-authored-by: Progi1984 <progi1984@gmail.com>

JevgenijVisockij committed:
Fixed test

JevgenijVisockij committed:
Added check for duplicate order names

JevgenijVisockij committed:
Adding an if for PHP Stan to shut it

JevgenijVisockij committed:
Fixing tests

matks commented:
Tested myself, I validate this is ✅  to be tested by QA. I also extended the "How to test".

khouloudbelguith commented:
Hi @JevgenijVisockij,

I have two issues:

**Issue1**
1. Go to BO > Shop Parameters > Order Settings > Statuses page
2. Create two order return statuses with the same name 
![image](https://user-images.githubusercontent.com/16067358/128518883-6b437ef0-a398-4aee-88b3-a3e21bba1ebf.png)
3. Go to `admin-dev/index.php/sell/customer-service/merchandise-return/` which is the list. Click on any of the existing returns to open the page.
4. See error 
![image](https://user-images.githubusercontent.com/16067358/128519002-a9e05ac8-7fe4-4bcc-8ec3-fec82bf4bed9.png)

**Issue2**
I think it is not related to this PR
1. Go to BO > Shop Parameters > Order Settings > Statuses page
2. Create an order return status with no color 
![image](https://user-images.githubusercontent.com/16067358/128519133-2513152f-ded8-44c7-b667-a8593cccb6e7.png)
3. Go to `admin-dev/index.php/sell/customer-service/merchandise-return/` which is the list. Click on any of the existing returns to open the page.
4. Change status to the last order return status created
5. See error 
![image](https://user-images.githubusercontent.com/16067358/128519280-34421a76-7902-4f3a-854f-48c894fb0c92.png)

Thanks!

matks commented:
Hello @khouloudbelguith why do you create two order return statuses with the same name? 😅 it is a very weird idea, isn't it ? A merchant is not supposed to do that?

JevgenijVisockij committed:
Adding ID in front of order return state to avoid duplicate order names issues

JevgenijVisockij committed:
Using sprintf

JevgenijVisockij commented:
> Hi @JevgenijVisockij,
> 
> I have two issues:
> 
> **Issue1**
> 
>     1. Go to BO > Shop Parameters > Order Settings > Statuses page
> 
>     2. Create two order return statuses with the same name
>        ![image](https://user-images.githubusercontent.com/16067358/128518883-6b437ef0-a398-4aee-88b3-a3e21bba1ebf.png)
> 
>     3. Go to `admin-dev/index.php/sell/customer-service/merchandise-return/` which is the list. Click on any of the existing returns to open the page.
> 
>     4. See error
>        ![image](https://user-images.githubusercontent.com/16067358/128519002-a9e05ac8-7fe4-4bcc-8ec3-fec82bf4bed9.png)
> 
> 
> **Issue2**
> I think it is not related to this PR
> 
>     1. Go to BO > Shop Parameters > Order Settings > Statuses page
> 
>     2. Create an order return status with no color
>        ![image](https://user-images.githubusercontent.com/16067358/128519133-2513152f-ded8-44c7-b667-a8593cccb6e7.png)
> 
>     3. Go to `admin-dev/index.php/sell/customer-service/merchandise-return/` which is the list. Click on any of the existing returns to open the page.
> 
>     4. Change status to the last order return status created
> 
>     5. See error
>        ![image](https://user-images.githubusercontent.com/16067358/128519280-34421a76-7902-4f3a-854f-48c894fb0c92.png)
> 
> 
> Thanks!

Fixed the first issue. 

Regarding second issue I could fix specifically for merchandise return, but it appears in more places than this. As far as I tested order return status list, order status list, order page and merchandise return list(But not order list) have the same issue. So I think it should be treated as a separate issue.

zuk3975 commented on src/Adapter/OrderReturn/QueryHandler/GetOrderReturnForEditingHandler.php:
```diff
@@ -0,0 +1,92 @@
+     */
+    private $orderRepository;
+
+    public function __construct(
```
missing docblock :see_no_evil:

zuk3975 commented on src/PrestaShopBundle/Resources/config/routing/admin/sell/customer_service/merchandise_return.yml:
```diff
@@ -23,3 +23,15 @@ admin_merchandise_returns_filter:
+    defaults:
+        _controller: PrestaShopBundle\Controller\Admin\Sell\CustomerService\MerchandiseReturnController::editAction
+        _legacy_controller: AdminReturn
+    #    _legacy_link: AdminReturn:updateorder_return
```
could add a todo reminder at the top of yml file, as I suppose these legacy links should be uncommented once the page is fully shipped to production?

zuk3975 commented on src/PrestaShopBundle/Resources/config/services/adapter/order_return.yml:
```diff
@@ -0,0 +1,31 @@
+services:
```
check whole file. Latest convention as I remember was to use 4 spaces indent in yml, it seems you are mixing 2 and 4

JevgenijVisockij committed:
Adding docblock

zuk3975 commented on tests/Integration/Adapter/OrderReturn/Repository/OrderReturnRepositoryTest.php:
```diff
@@ -0,0 +1,55 @@
+        $this->orderReturnRepository = self::$container->get('prestashop.adapter.order_return.repository.order_return_repository');
+    }
+
+    public function testGetThrowsException()
```
return type

zuk3975 commented on tests/Integration/Adapter/OrderReturnState/Repository/OrderReturnStateRepositoryTest.php:
```diff
@@ -0,0 +1,55 @@
+        $this->orderReturnStateRepository = self::$container->get('prestashop.adapter.order_return_state.repository.order_return_state_repository');
+    }
+
+    public function testGetThrowsException()
```
return type

JevgenijVisockij committed:
Fixes to yml files

JevgenijVisockij committed:
Adding missing return type

JevgenijVisockij committed:
Adding missing return type

matks commented on src/Adapter/OrderReturn/Repository/OrderReturnRepository.php:
```diff
@@ -0,0 +1,84 @@
+        /* @var OrderReturn $orderReturn */
+        $orderReturn = $this->getObjectModel($orderReturnId->getValue(), OrderReturn::class, OrderReturnNotFoundException::class);
+        if (!$orderReturn instanceof OrderReturn) {
+            throw new OrderReturnException('Must be instance of OrderReturn');
```
Nice-to-have: print full fully qualified name
```suggestion
            throw new OrderReturnException(sprintf('Must be instance of %s', OrderReturn::class);
```

matks commented on src/Adapter/Form/ChoiceProvider/OrderReturnStateChoiceProvider.php:
```diff
@@ -0,0 +1,69 @@
+        $orderStates = OrderReturnState::getOrderReturnStates($this->contextLangId);
+
+        foreach ($orderStates as $orderState) {
+            $choices[sprintf('%s - %s', $orderState['id_order_return_state'], $orderState['name'])]
```
Question: is `$orderState['name']` translated?
If the order state has a different name in Lithuanian and English, then whether I use EN or LT as back office language, I will be displayed the right value for this order state ?

JevgenijVisockij replied:
Yes OrderReturnStates takes names correctly according to language you pass(which is contextLangId).

matks reviewed (dismissed):
Code is good for me (thanks @zuk3975 for the review !)
I added one nice-to-have and one question.

JevgenijVisockij committed:
Update src/Adapter/OrderReturn/Repository/OrderReturnRepository.php

IMO makes sense.

Co-authored-by: Mathieu Ferment <mathieu.ferment@prestashop.com>

JevgenijVisockij committed:
Added missing )

Progi1984 commented on src/PrestaShopBundle/Resources/config/services/adapter/order_return.yml:
```diff
@@ -0,0 +1,31 @@
+       - '@prestashop.adapter.customer.repository.customer_repository'
+       - '@prestashop.adapter.order.repository.order_repository'
+
+  prestashop.adapter.order_return.command_handler.edit_order_return_state_handler:
```
```suggestion
  prestashop.adapter.order_return.command_handler.update_order_return_state_handler:
```

Progi1984 commented on src/Adapter/Form/ChoiceProvider/OrderReturnStateChoiceProvider.php:
```diff
@@ -0,0 +1,69 @@
+use PrestaShop\PrestaShop\Core\Domain\OrderReturnState\Exception\OrderReturnStateException;
+use PrestaShop\PrestaShop\Core\Form\FormChoiceProviderInterface;
+
+final class OrderReturnStateChoiceProvider implements FormChoiceProviderInterface
```
```suggestion
class OrderReturnStateChoiceProvider implements FormChoiceProviderInterface
```

Progi1984 reviewed (commented):
@JevgenijVisockij Thank you for your contribution. I have added  two last comments before approving your PR.

matks committed:
Apply suggestions from code review

Co-authored-by: Progi1984 <progi1984@gmail.com>

matks reviewed (approved):
Great !

khouloudbelguith commented:
Hi @JevgenijVisockij,

Thanks to check the Conflicting files.

Thank you!

PierreRambaud commented:
Hi @JevgenijVisockij ,

Since we had no news from you for more than 30 days, I'll close this pull request. Feel free to reopen or open another one if you think it's still relevant.

Thanks!
